storage/src/main/scala/coop/rchain/storage/regex/Fsm.scala

```diff
@@ -0,0 +1,811 @@
+    val pendingFinals = Queue[String]()
+    var needInitial = true
+
+    def nextStr: Option[String] =
+      if (pendingFinals.nonEmpty) {
+        Some(pendingFinals.dequeue())
+      } else {
+        //Initial entry (or possibly not, in which case this is a short one)
+        if (needInitial) {
+          needInitial = false
+          val cstate = this.initialState
+          if (livestates.contains(cstate)) {
+            pending += "" -> cstate
+            if (finalStates.contains(cstate)) {
+              pendingFinals += ""
+            }
+          }
+        }
+
+        while (pendingFinals.isEmpty && pending.nonEmpty) {
+          val (pendingString, pendingState) = pending.dequeue()
+
+          val transition = transitions.get(pendingState)
+          if (transition.isDefined) {
+            for ((nextSymbol, nextState) <- transition.get) {
+              val nextString = pendingString + nextSymbol
+
+              if (livestates.contains(nextState)) {
+                pending += nextString -> nextState
+                if (finalStates.contains(nextState)) {
+                  pendingFinals += nextString
+                }
+              }
+            }
+          }
+        }
+
+        if (pendingFinals.nonEmpty) {
+          Some(pendingFinals.dequeue())
+        } else {
+          None
+        }
+      }
+
+    def infStream: Stream[String] = {
+      val str = nextStr
+      if (str.isDefined) {
+        str.get #:: infStream
+      } else {
+        Stream.empty[String]
+      }
+    }
+
+    infStream
+  }
+
+  /**
+    * Given an FSM and a multiplier, return the multiplied FSM.
+    */
+  def Times(multiplier: Int): Fsm = {
+    require(multiplier >= 0, "Can't multiply an FSM by " + multiplier)
+    //here we always work with "currentState -> iteration" pairs
+    def initial = HashSet(this.initialState -> 0)
+
+    def follow(crawlState: HashSet[(Int, Int)],
+               symbol: Char): Option[HashSet[(Int, Int)]] = {
+      val next = HashSet[(Int, Int)]()
+
+      crawlState.foreach(fsmStateAndIteration => {
+        val fsmState = fsmStateAndIteration._1
+        val iteration = fsmStateAndIteration._2
+
+        if (iteration < multiplier) {
+          val transition = transitions.get(fsmState)
+          if (transition.nonEmpty) {
+            val subState = transition.get.get(symbol)
+            if (subState.nonEmpty) {
+              next += subState.get -> iteration
+              //final of self? merge with initial on next iteration
+              if (finalStates.contains(subState.get)) {
+                next += initialState -> (iteration + 1)
+              }
+            }
+          }
+        }
+      })
+
+      if (next.nonEmpty) {
+        Some(next)
+      } else {
+        None
+      }
+    }
+
+    //If the initial state is final then multiplying doesn't alter that
+    def isFinal(crawlState: HashSet[(Int, Int)]): Boolean =
+      crawlState.exists(fsmStateAndIteration =>
```

Comment:
You can use destructuring here:
`crawlState.exists { (first,second) => ...`

Reply:
Done